External/FEXCore/Source/Interface/HLE/Thunks/Thunks.cpp

```diff
@@ -151,15 +169,89 @@ namespace FEXCore {
         uint8_t *HostTrampolineInstanceDataPtr;
         size_t HostTrampolineInstanceDataAvailable = 0;
 
+        std::unordered_map<unsigned, AsyncCallbackState> AsyncWorkerThreads;
+        std::mutex AsyncWorkerThreadsMutex;
+
+        /**
+         * Registers the calling thread as a worker thread for callback
+         * functions that are asynchronously invoked in a host context.
+         *
+         * Such a worker thread must be designated since otherwise there is no
+         * x86 context to run the callback in. Most importantly, this would
+         * prevent TLS from working.
+         *
+         * Before the worker thread shuts down, UnregisterAsyncWorkerThread
+         * must be called.
+         */
+        static void RegisterAsyncWorkerThread(void* argsv) {
+          struct args_t {
+              unsigned id;
+          } args = *reinterpret_cast<args_t*>(argsv);
+
+          auto ThunkHandler = reinterpret_cast<ThunkHandler_impl*>(Thread->CTX->ThunkHandler.get());
+
+          AsyncCallbackState *WorkerState = nullptr;
+          {
+            // Create and initialize new entry
+            std::unique_lock lock(ThunkHandler->AsyncWorkerThreadsMutex);
+            WorkerState = &ThunkHandler->AsyncWorkerThreads[args.id];
+            WorkerState->Thread = Thread;
```

Comment:
I think this should probably modify the thread state coming in to this to modify RSP to offset by the redzone.